File: Caller/callback.cpp
#include "callback.h"
#include <iostream>

void Caller::call(Callback* cb) {
  if (cb) {
    std::cout << "C++ Caller::call: invoking run()..." << std::endl;
    cb->run(); // This will call the JS version if cb points to a JS object
    std::cout << "C++ Caller::call: run() finished." << std::endl;
  } else {
    std::cout << "C++ Caller::call: callback object is null." << std::endl;
  }
}

File: Caller/callback.js
const addon = require("./build/Release/Callback");

class CustomCallback extends addon.Callback {
  constructor() {
    super();
    console.log("JS: CustomCallback constructor");
  }
  
  run() {
    console.log("JS: CustomCallback run method called");
  }
}

console.log("JS: Creating Caller");
const caller = new addon.Caller();
const customCallback = new CustomCallback();

caller.call(customCallback);

